presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java

```diff
@@ -62,4 +64,23 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
                 parent.replaceChildren(
                         child.getSources())));
     }
+
+    /**
+     * @return an identity projection containing only those node outputs also in permittedOutputs, or the original node if all its inputs were permitted.
+     */
+    public static PlanNode restrictOutputs(PlanNodeIdAllocator idAllocator, PlanNode node, Set<Symbol> permittedOutputs)
+    {
+        List<Symbol> restrictedOutputs = node.getOutputSymbols().stream()
+                .filter(permittedOutputs::contains)
+                .collect(toImmutableList());
+
+        if (restrictedOutputs.size() < node.getOutputSymbols().size()) {
+            return new ProjectNode(
+                    idAllocator.getNextId(),
+                    node,
+                    Assignments.identity(restrictedOutputs)
```

Comment:
It's not a correctness problem.  The former PlanNode is still there, now as a child.  So the narrower ProjectNode would be the parent of the original PlanNode, in your example.  Rewriting the original ProjectNode is what the output-pruning optimization does.  In combination with RemoveRedundantIdentityProjections, the plan will eventually become what you want.